### PR TITLE
[FIX] website: confirmation modal on discard

### PR DESCRIPTION
**Pull request by dhba-odoo**
**Current behavior before PR:**

When entering edit mode on the website and discarding directly, the editor
considers that there are changes in the DOM and displays a modal with
"Are you sure you want to discard because there are pending changes".

**Desired behavior after PR is merged:**

Now, when entering edit mode on the website and discarding it directly will not
display a confirmation modal.

Task-3056463

### CODE REVIEW

File: addons/website/static/src/js/tours/tour_utils.js [hunk 1]
```diff
@@ -221,6 +235,14 @@ function clickOnText(snippet, element, position = "bottom") {
     };
 }
 
+function clickOnWebsite(position = "bottom") {
+    return {
+        trigger: ".o_app[data-menu-xmlid='website.menu_website_configuration']",
+        content: Markup(_t("<b>Click Website Icon</b> to Open Website.")),
+        position: position,
+    };
+}
```
Comment (dmo-odoo) [inline addons/website/static/src/js/tours/tour_utils.js:244]: I'm surprised. These are the website tours. I would have expected them to already have some way to open the website. Why do we need a new one ?

File: addons/website/static/src/js/tours/tour_utils.js [hunk 2]
```diff
@@ -139,6 +139,20 @@ function changePaddingSize(direction) {
...
+        extra_trigger: "body:not(:has(.o_dialog))",
+        content: Markup(_t("<b>Click Discard</b> to Discard all Changes.")),
+        position: position,
+    }, {
+        trigger: "iframe body:not(.editor_enable)",
+        noPrepend: true,
+        auto: true,
+        run: () => null,
+    }];
+}
```
Comment (dmo-odoo) [inline addons/website/static/src/js/tours/tour_utils.js:154]: Don't they already have tours when they use the Discard button ? I'm surprised we need a new one.

File: addons/website/static/src/js/content/snippets.animation.js [hunk 3]
```diff
@@ -1449,7 +1449,13 @@ registry.ImagesLazyLoading = publicWidget.Widget.extend({
...
         const imgEls = this.$target[0].querySelectorAll('img[data-lazy-loading-initial-min-height]');
         for (const imgEl of imgEls) {
-            this._restoreImage(imgEl);
+            imgEl.style.minHeight = '1px';
+            wUtils.onceAllImagesLoaded($(imgEl)).then(() => {
+                if (this.isDestroyed()) {
+                    return;
+                }
+                this._restoreImage(imgEl);
+            });
```
Comment (dmo-odoo) [inline addons/website/static/src/js/content/snippets.animation.js:1458]: I do not understand why this change is needed here.

File: addons/website/static/src/components/wysiwyg_adapter/wysiwyg_adapter.js [hunk 4]
```diff
@@ -226,7 +226,8 @@ export class WysiwygAdapterComponent extends ComponentAdapter {
             this.props.quitCallback({ onLeave, reloadIframe });
         };
 
-        if (!forceLeave && this._isDirty()) {
+        const isDirty = this._isDirty() && this.$editable[0].ownerDocument.defaultView.$('.o_dirty').length;
```
Comment (dmo-odoo) [inline addons/website/static/src/components/wysiwyg_adapter/wysiwyg_adapter.js:229]: This is not correct IMO. If the `_isDirty` method returns `true` when we expect it to return `false` then it is that method that should be fixed, not just adding another condition on top of it.

File: addons/website/static/src/js/tours/tour_utils.js (hunk 2 above)
Comment (hbh-odoo) [reply addons/website/static/src/js/tours/tour_utils.js:154]: @dmo-odoo Yes! need to add a new one cause there is no tour yet about `Discard`

File: addons/website/static/src/js/content/snippets.animation.js (hunk 3 above)
Comment (dhba-odoo) [reply addons/website/static/src/js/content/snippets.animation.js:1458]: @dmo-odoo  When we start editing the website if the image is not loaded properly then it will contain the `data-lazy-loading-initial-min-height` attribute that is removed from the ` _restoreImage` method. Here method ` _restoreImage` does not call for the specific images. If we don't stop the removal of that attribute from the images which is not loaded yet then at the discard time this small change is taken as the dom has changed.
Hence, the processing is done when the image is loaded successfully

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 5]
```diff
@@ -839,7 +839,8 @@ const Wysiwyg = Widget.extend({
      * @returns {Boolean}
      */
     isDirty: function () {
-        return this._initialValue !== (this.getValue() || this.$editable.val());
+        const isDocumentDirty = this.$editable[0].ownerDocument.defaultView.$(".o_dirty").length;
+        return this._initialValue !== (this.getValue() || this.$editable.val()) && isDocumentDirty;
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:843]: When `this._initialValue !== (this.getValue() || this.$editable.val())` is `true` but `isDocumentDirty` is `false`, what are the differences between `this._initialValue` and `(this.getValue() || this.$editable.val())` ? There must be a difference, otherwise there would be not bug to fix here, so could you tell me what the difference is exactly ?

File: addons/website/static/src/js/tours/homepage_edit_discard.js [hunk 6]
```diff
@@ -0,0 +1,6 @@
+/** @odoo-module alias=website.homepage_edit_discard*/
+"use strict";
+
+import wTourUtils from "website.tour_utils";
+
+wTourUtils.homepageEditDiscard("homepage_edit_discard", []);
```
Comment (dmo-odoo) [inline addons/website/static/src/js/tours/homepage_edit_discard.js:6]: Do we really want an entire file for this small tour ? Nothing wrong with that, I'm just wondering. @qsm-odoo What is your policy for tours ? Like one tour per file ?

Comment (qsm-odoo) [reply addons/website/static/src/js/tours/homepage_edit_discard.js:6]: There is no fixed rule. In general, yes one tour per file I guess. But if they are very related or even part of the same python tour, of course makes sense to have multiples in the same file. But in the case of this PR @dhba-odoo:

- There are issues with the tour system at the moment (maybe this affects this) but `saveAs` is meant to override an existing tour, which I don't think is what you are doing here ?

- Why do we need an util for creating tours that click on the discard button at the end? Seems overkill.

- Don't you mean to create a python test out of this? And is not this supposed to be a test tour (not shown to users) (in which case, missing `test: true` parameter here + tour defined in the wrong folder ?)

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 5 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:843]: @dmo-odoo There are some cases in which the DOM is changed even though we don't perform any operation.
Like, 1) If there are multiple menus available on the website, and we start editing the website some menus will convert into dropdowns.
2) Only scrolling up and down on the website will sometimes add or remove tags and attributes.
At a time like this value of this._initialvalue and (this.getValue() || this.$editable.val()) will not match. So, for those cases, I have checked isDocumentDirty.

Comment (dmo-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:843]: @Goaman I'm not very fond of the proposed fix here. Do you think we could filter out the changes mentionned by @dhba-odoo instead ? I would prefer it but in a way it might be very hard (or maybe even impossible) to list all possible cases to ignore. At least the proposed fix has the merit of being more generic, in a way. What do you think ?

File: addons/website/static/src/js/content/snippets.animation.js (hunk 3 above)
Comment (dmo-odoo) [reply addons/website/static/src/js/content/snippets.animation.js:1458]: @Goaman Just fyi, another case here that can trigger unwanted changes.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 5 above)
Comment (Goaman) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:848]: @dmo-odoo It is not the first time I see the value of the editable `(this.getValue() || this.$editable.val())` being different even without any modification. At some point, it happened with the checklist in `html_field` (I'm not sure if it still does but anyway, html_field has a custom dirty check). 
For me, we should *not* rely on `(this.getValue() || this.$editable.val())`.
Checking for the class "o_dirty" in website seems a more robust approach.

File: addons/website/static/tests/tours/homepage_edit_discard.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+/** @odoo-module alias=website.homepage_edit_discard*/
```
Comment (Goaman) [inline addons/website/static/tests/tours/homepage_edit_discard.js:1]: Why do you need an alias?

File: addons/website/static/src/js/content/snippets.animation.js (hunk 3 above)
Comment (Goaman) [reply addons/website/static/src/js/content/snippets.animation.js:1458]: As it is an attribute, we can add a rule in `filterMutationRecords` of `wysiwyg.js` like:
`!(record.type === 'attributes' && record.name === 'data-lazy-loading-initial-min-height')`

Comment (dhba-odoo) [reply addons/website/static/src/js/content/snippets.animation.js:1458]: @Goaman I tried it but it's not working.

Comment (Goaman) [reply addons/website/static/src/js/content/snippets.animation.js:1458]: Why did it not worked?

Comment (dhba-odoo) [reply addons/website/static/src/js/content/snippets.animation.js:1458]: @Goaman 
Here, when I click on the edit button 'data-lazy-loading-initial-min-height' is removed from the element, and in return because of that o_dirty class is added to that particular element. 

Even if I try your suggestion o_dirty class is still added to that Element and because of that, my other condition is not working correctly when I try to check if there is any o_dirty class available in dom.

File: addons/website/static/tests/tours/homepage_edit_discard.js (hunk 7 above)
Comment (dhba-odoo) [reply addons/website/static/tests/tours/homepage_edit_discard.js:1]: @Goaman
I have observed some test tours in which it's used. There is no particular reason for this. 
If it's not needed I will remove it.

Comment (Goaman) [reply addons/website/static/tests/tours/homepage_edit_discard.js:1]: Indeed it is not needed.

File: addons/website/static/src/js/content/snippets.animation.js (hunk 3 above)
Comment (Goaman) [reply addons/website/static/src/js/content/snippets.animation.js:1458]: The method that add the `o_dirty` class should be `processRecords` of `_setObserver`. 
It only add the class on target of a mutation for the record that are not filtered.
My question is: if the record is properly filtered, what part of the code is adding the `o_dirty` class ?

Comment (dhba-odoo) [reply addons/website/static/src/js/content/snippets.animation.js]: @Goaman  I have found the reason o_dirty class was added even after the record is properly filtered. the reason for that is that when 'data-lazy-loading-initial-min-height' is removed from the element at the same time min-height style is also removed from that element. I have added record.attributeName === 'style' in the above rule and it's working.
I have already push the changes in the branch.

File: addons/website/static/tests/tours/homepage_edit_discard.js (hunk 7 above)
Comment (dhba-odoo) [reply addons/website/static/tests/tours/homepage_edit_discard.js]: @Goaman Done!

File: addons/website/static/src/js/content/snippets.animation.js (hunk 3 above)
Comment (dmo-odoo) [reply addons/website/static/src/js/content/snippets.animation.js]: Hello @dhba-odoo, we have investigated the issue with @Goaman and found the root cause. The dirty observer was called before the public widgets were stopped and therefore it observed the changes caused by the destroying of the public widgets. I have updated the commit accordingly. @detrouxdev Could you validate our change here please ? In case there was a reason for calling `widgets_start_request` after `_setObserver` that we are not aware of ? cc @qsm-odoo

Comment (detrouxdev) [reply addons/website/static/src/js/content/snippets.animation.js]: Good catch 👍🏻  but I think that the issue might also be present in 15.0

Around this line (15.0):
https://github.com/odoo/odoo/blob/797f67bcb222142b79a9eee3e4a882ff679af9db/addons/website/static/src/js/menu/edit.js#L240

We pretty much replaced `_createWysiwyg` into an OWL `useEffect`, as the Wysiwyg is the main widget of this Component, it's started with the OWL lifecycle. When we did so, we moved what was done after `await this.wysiwyg.attachTo($('#wrapwrap'));` into this useEffect, while moving the code that set the observer into its own method, (I do not clearly remember why).

I just did a quick test and it seems like this could be the case on 15.0 as well:
- Drop a product snippet
- Save
- Enter edit mode
- Discard
- => Discard Popup when we should not have one (?)

So my opinion would be that I do not mind the fix, but it should probably be done in 15.0. And I am not familiar with what the OdooEditor observes during that time but maybe it should also not listen to changes during that time? You're probably better at judging that than me though ^^

Comment (dmo-odoo) [reply addons/website/static/src/js/content/snippets.animation.js]: @detrouxdev Thank you for your input ! Since the changes in 15.0 would not forward port properly to 16.0 anyway, we'll be moving forward with this PR but I'm going to create a new task for fixing this issue in 15.0 as well with all the details you provided. Thanks !

**Comment by dmo-odoo** [comment]
@robodoo r+
